Use a custom deserializer instead of wrapper types

romac opened the pull request.

romac committed:
Add `serde-itf` crate with native deserialization

romac committed:
Use the Visitor pattern

romac committed:
Support for BigInt

romac committed:
Move values under `value` module

romac committed:
Cleanup

romac committed:
Move tests into `serde-itf` crate and delete `itf` crate

romac committed:
Rename `serde-itf` crate to `itf`

rnbguy committed:
update visit_bigint for num_bigint::BigInt

rnbguy committed:
update itf::value::BigInt deserialize for num_bigint::BigInt

rnbguy committed:
visit_bigint when deserialize_tuple

rnbguy committed:
all integers are bigints

rnbguy committed:
use num_bigint::BigInt in deserialized stuct directly

rnbguy committed:
reactivate failing test

rnbguy committed:
rm unused import

rnbguy committed:
fix tuple deserialize

rnbguy committed:
fix set deserialize

rnbguy committed:
regression tests

rnbguy commented:
To document how I solved the `num_bigin::BigInt` deserialization.

The ITF trace deserialization to native types works as follows,
1. deserialize raw `itf` JSON to `serde_json::Value`
2. then deserializing `serde_json::Value` to `itf::value::Value` (`impl<'de> Deserialize<'de> for itf::value::Value`) 
3. then deserializing `itf::value::Value` to `T` (`impl<'de> Deserializer<'de> for itf::value::Value`)

To get native deserialization of `num_bigint::BigInt`, `itf::value::Value::BigInt` needs to be deserialized as `(int, [uint])`. I added https://github.com/informalsystems/itf-rs/pull/6/commits/a56363c5e86d1cf30bd3c6aa651914cc289ad13b for this. And, https://github.com/informalsystems/itf-rs/pull/6/commits/92bb9d3d12adf7e4b26b0119385d8f5b2415fcbc allows `itf::value::Value::BigInt` to be deserialized as a tuple.

But this creates an weird issue - `itf::from_str::<itf::value::Value>` will deserialize `{"#bigint": "99"}` as `[1, [99]]`. As `itf::from_str` will deserialize itf-bigint format to `itf::value::Value` and then again to itself. The second deserialization makes it a tuple and `itf::value::Value` does not have any information on how to deserialize it back to `itf::value::Value::BigInt`. So I added https://github.com/informalsystems/itf-rs/pull/6/commits/c03bec1122fbb533ffe19e6c5dcac1c3465cfb82.

Note, with this, `itf::from_str` does not differentiate between `{"#bigint": "99"}` and `[1, [99]]`; they both can be deserialized to `num_bigint::BigInt`.

rnbguy commented:
Note, this does not support `num_bigint::BigUint` which has a different deserialization format than `num_bigint::BigInt`.

- `BigInt`: `(int, [uint])`
- `BigUint`: `[uint]`

rnbguy committed:
bunch of bigint testcases from debugging

rnbguy committed:
rm dbg!()

rnbguy committed:
use serde_json::Value over json string

rnbguy committed:
add testcases for Value::deserialize(Value)

rnbguy commented:
`Value::deserialize(Value)` is creating equivalency issues for `Value::{Set, Map, List}` too. I added testcases in https://github.com/informalsystems/itf-rs/pull/6/commits/d89d2fb5133b61e18d33a62971168e8ecd0fa797.

I suggest making `itf::Value` private. This way we can also revert https://github.com/informalsystems/itf-rs/commit/c03bec1122fbb533ffe19e6c5dcac1c3465cfb82 :slightly_smiling_face:

romac commented:
Amazing stuff @rnbguy! Thanks so much!

> I suggest making itf::Value private. This way we can also revert https://github.com/informalsystems/itf-rs/commit/c03bec1122fbb533ffe19e6c5dcac1c3465cfb82 🙂

Yeah let's do that!

romac committed:
Merge branch 'main' into native-types

romac committed:
Update code coverage workflow

romac committed:
Update codecov config

romac committed:
Hide `Value` enum from documentation

We keep exporting it to allow for usecases where one needs the raw ITF `Value`,
but avoid documenting it as there are a few pitfalls when doing so.

See <https://github.com/informalsystems/itf-rs/pull/6#issuecomment-1817860601> for more information

rnbguy committed:
revert c03bec1

rnbguy committed:
update tests

rnbguy committed:
fix failing tests

rnbguy committed:
fix incorrect test

rnbguy committed:
itf::Value over itf::value::Value

rnbguy committed:
add a complete test

rnbguy committed:
add portability test between i64 and BigInt

rnbguy committed:
support deserializing Number to BigInt

rnbguy committed:
test deserialization failure from true bigint to i64

romac commented on itf/tests/regression.rs:
```diff
@@ -124,3 +124,42 @@ fn test_map_with_non_str_key() {
 }
+
+#[test]
+fn test_complete() {
```
Nice!

romac commented on itf/src/de.rs:
```diff
@@ -258,6 +258,7 @@ impl<'de> Deserializer<'de> for Value {
     {
         match self {
             Value::BigInt(v) => visit_bigint(v, visitor),
+            Value::Number(v) => visit_bigint(BigInt::new(v), visitor),
```
Good idea!

rnbguy committed:
add success case for true bigint deserialization

rnbguy committed:
test for failure when deserialize bigint to int in enum

romac committed:
Add example and update README

romac committed:
Update changelog

rnbguy commented:
@romac any idea why this (https://github.com/informalsystems/itf-rs/pull/6/commits/9920493b3a74865f1d41aca72838a6e57d48d9b5) is not working?

https://github.com/informalsystems/itf-rs/blob/9920493b3a74865f1d41aca72838a6e57d48d9b5/itf/tests/regression.rs#L141-L154

romac committed:
Include README in top-level crate doc

romac committed:
Split `Deserializer` impl and `Error` into their own modules

romac committed:
Add a `itf::de::from_bigint` deserializer function to convert from `BigInt` via `#[serde(deserialize_with = "...")]`

romac commented:
Ah good catch! Yeah it's a shame we can't seem to support this.

That said, we can perhaps still enable this by providing a `from_bigint<'de, A: TryFrom<BigInt>, D>(d: D) -> Result<A, D::Error>` helper function which could be used with `#[serde(deserialize_with = "itf::de::from_bigint")]`.

See https://github.com/informalsystems/itf-rs/pull/6/commits/a2eef53c28f230d1ec0ec24578815f7b2ab3b132

rnbguy commented:
Yea `serde(with)` is the last straw :sweat_smile: 

Btw. I would prefer `serde_with::DeserializeAs` with `serde_with::As`. This adds a dependency to `serde_with`. But this way, we can support container types too.

```rs
use itf::value::BigInt;
use serde_with::As;

impl<'de, T> serde_with::DeserializeAs<'de, T> for BigInt { ... }

#[derive(Deserialize, Debug)]
#[serde(tag = "typ")]
enum FooBarWithInt {
    Foo {
        #[serde(with = "As::<Vec<Vec<BigInt>>>")]
        _foo: Vec<Vec<i64>>,
    },
    Bar {
        _bar: String,
    },
}
```

I can take care of the changes tomorrow :slightly_smiling_face:

romac commented:
Ah good point, didn't think of containers! I am fine with adding `serde-with` as a dep :)

rnbguy committed:
disallow implicit deser from bigint to (u)int

rnbguy committed:
disallow implicit deser from int to bigint

rnbguy committed:
use Integer as DeserializeAs to deserialize from BigInt to (u)int

rnbguy committed:
update tests for explicit serialization from BigInt to (u)int

rnbguy commented:
I removed the code that implicitly serialized `BigInt` to `(u)int` and vice-versa. Now, `{"#bigint": "<num>"}` is always deserialized to `BigInt`.

For rust primitive integer types, a user has to pass `serde_with::As` type in _[serde-with](https://serde.rs/field-attrs.html#with)_ (e.g. `#[serde(with = "As::<itf::de::Integer>")]`) or `serde_with::serde_as` attribute accordingly.

rnbguy commented:
Btw, `serde_with` provides `TryFromInto` - which can be used as follows - `#[serde(with = "As::<TryFromInt<BigInt>>")]` in our context. `itf::de::Integer` is just a shorthand for it.

romac commented:
Awesome!

romac committed:
Re-export `serde_with::As` under `itf::de::As` to enable its use without explicitly depending on `serde-with`

rnbguy commented:
Btw, maybe it make sense to redefine `Integer` as just

```rs
pub type Integer = TryFromInto<BigInt>;
```

romac committed:
Add doc comments to the helpers

romac committed:
Document public items

romac committed:
Remove unused `Error::Io` variant

romac committed:
Document `Error`

romac committed:
Redefine `itf::de::Integer` as `TryFromInto<BigInt>`
